### DISCUSSION TITLE
How to useReactFlow within zustand actions?

### DISCUSSION THREAD

**Discussion by tauren** [xyflow/xyflow#3861] [Q&A] [answered]
I would like to use functions from `useReactFlow` within my *zustand* store but I'm getting errors when trying. Is this possible, and if so how?

My app allows nodes to be created by dragging a div from a toolbar onto the flow. The `onDrop` handler below works fine when defined in the React component that renders ReactFlow. But when I move it to my store, it breaks on this line:

`const { addNodes, screenToFlowPosition } = useReactFlow();`

Source below is from the [State Management](https://reactflow.dev/learn/advanced-use/state-management) example, with `onDrop` added.

```
const useStore = create<RFState>((set, get) => ({
  nodes: initialNodes,
  edges: initialEdges,
  onNodesChange: (changes: NodeChange[]) => {
    set({
      nodes: applyNodeChanges(changes, get().nodes),
    });
  },
  onEdgesChange: (changes: EdgeChange[]) => {
    set({
      edges: applyEdgeChanges(changes, get().edges),
    });
  },
  onConnect: (connection: Connection) => {
    set({
      edges: addEdge(connection, get().edges),
    });
  },
    onDrop: (event) => {
      set(() => {
        event.preventDefault();
        const type = event.dataTransfer.getData('application/reactflow');
        if (typeof type === 'undefined' || !type) return {};
        const data = getNodeTemplate(type);

        // FIXME: This doesn't work here
        const { addNodes, screenToFlowPosition } = useReactFlow();

        data.position = screenToFlowPosition({
            x: event.clientX,
            y: event.clientY,
        });
        addNodes(data);
        return {};
        });
    },
},

}));
```

**Comment by bcakmakoglu** [comment, marked as answer]
Instead of putting `onDrop` into your store, put it into a component that has access to the `ReactFlowProvider` context.
You also don't need to use `addNodes` of `useReactFlow` - you already have access to your nodes in your store, so just add it in there 😄 
Using `addNodes` would just be a roundabout way of doing the same in this case (unless you explicitly want to track changes for undo/redo) as `addNodes` would create a new change, emit it through `onNodesChange` which then would update your store with the new node.
